src/resources/2019/DiabolicPizza.ts

```diff
@@ -0,0 +1,92 @@
+import { autosellPrice, getWorkshed, toInt, visitUrl } from "kolmafia";
+import { get as getModifier } from "../../modifier";
+import { have as haveItem } from "../../lib";
+import { $item, $monster } from "../../template-string";
+import { clamp } from "../../utils";
+
+/**
+ * @returns true if the Diabolic pizza cube is installed in workshed.
+ */
+export function installed(): boolean {
+  return getWorkshed() === $item`diabolic pizza cube`;
+}
+
+/**
+ * @returns true if the Diabolic pizza cube is installed in workshed or in inventory.
+ */
+export function have(): boolean {
+  return installed() || haveItem($item`diabolic pizza cube`);
+}
+
+/**
+ * @param items One or more items to be checked.
+ * @returns true if all items are a valid pizza ingredients.
+ */
+export function validIngredients(...items: Item[]): boolean {
+  const valid = (i: Item): boolean => i.tradeable && i.discardable && !i.gift;
+  return items.every((item) => valid(item));
+}
+
+/**
+ * Simulates a pizza given a set of four ingredients. Ingredient order determines the possible effects.
+ * Possible effects depend on the KoLMafia nohookah flag and avatar potion data.
+ * @param a First pizza ingredient
+ * @param b Second pizza ingredient
+ * @param c Third pizza ingredient
+ * @param d Fourth pizza ingredient
+ * @returns Returns an object with properties:
+ *   adventures: the number of adventures this pizza would generate, before modifiers (3-15 advs).
+ *   duration: the duration of the effect of the pizza (5-100 advs).
+ *   effects: an array of possible effects that this pizza can yield.
+ * @returns -1 and an empty array if the ingredients are not valid.
+ */
+export function simulate(
+  a: Item,
+  b: Item,
+  c: Item,
+  d: Item
+): { adventures: number; effects: Effect[]; duration: number } {
+  const abcd = [a, b, c, d] as const;
+  const error = { adventures: -1, effects: [], duration: -1 };
+  if (!validIngredients(...abcd)) return error;
+  const totalLetters = abcd.reduce((sum, item) => sum + item.nameLength, 0);
+  const advs = clamp(Math.floor(totalLetters / 10), 3, 15);
```

Comment:
```suggestion
  const adventures = clamp(Math.floor(totalLetters / 10), 3, 15);
```
I prefer not abbreviating variable names unless absolutely necessary